src/DevTools/FileReflection.php

```diff
@@ -168,9 +178,18 @@ public function reset_theme_variables() {
 	 * }
 	 */
 	public function get_file_source( $file ) {
+		static $recursion_protection = false;
+
+		if ( $recursion_protection ) {
+			return [];
+		}
+
+		$recursion_protection = true;
```

Comment:
Most of this PR is not directly about fixing the issue, but about making the tests less flaky.

The above change is what actually fixes the issue. The first attempt at fixing this was about trying to avoid a condition where it would happen. But playing around with the tests and their global state just made the issue appear in other places, most notably in the validation manager, where the wrapped callback was requesting a file source. And it happened to wrap the very filters that were causing the loop.

Therefore, a better fix is to assume the possibility for a recursion loop is always present, and detect and break it instead, with a simple safeguard.

Reply:
For reference, the backtrace for when recursion is happening is:

```
#0  AmpProject\AmpWP\DevTools\FileReflection->get_file_source() called at [/app/public/content/plugins/amp/src/DevTools/CallbackReflection.php:68]
#1  AmpProject\AmpWP\DevTools\CallbackReflection->get_source() called at [/app/public/content/plugins/amp/includes/validation/class-amp-validation-manager.php:1325]
#2  AMP_Validation_Manager::wrap_hook_callbacks() called at [/app/public/core-dev/src/wp-includes/class-wp-hook.php:333]
#3  WP_Hook->do_all_hook() called at [/app/public/core-dev/src/wp-includes/plugin.php:875]
#4  _wp_call_all_hook() called at [/app/public/core-dev/src/wp-includes/plugin.php:189]
#5  apply_filters() called at [/app/public/core-dev/src/wp-includes/option.php:81]
#6  get_option() called at [/app/public/core-dev/src/wp-includes/theme.php:315]
#7  get_template() called at [/app/public/core-dev/src/wp-includes/theme.php:326]
#8  get_template_directory() called at [/app/public/content/plugins/amp/src/DevTools/FileReflection.php:254]
#9  AmpProject\AmpWP\DevTools\FileReflection->get_template_directory() called at [/app/public/content/plugins/amp/src/DevTools/FileReflection.php:336]
#10 AmpProject\AmpWP\DevTools\FileReflection->is_parent_theme_file() called at [/app/public/content/plugins/amp/src/DevTools/FileReflection.php:197]
#11 AmpProject\AmpWP\DevTools\FileReflection->get_file_source() called at [/app/public/content/plugins/amp/src/DevTools/CallbackReflection.php:68]
#12 AmpProject\AmpWP\DevTools\CallbackReflection->get_source() called at [/app/public/content/plugins/amp/includes/validation/class-amp-validation-manager.php:1325]
#13 AMP_Validation_Manager::wrap_hook_callbacks() called at [/app/public/core-dev/src/wp-includes/class-wp-hook.php:333]
#14 WP_Hook->do_all_hook() called at [/app/public/core-dev/src/wp-includes/plugin.php:875]
#15 _wp_call_all_hook() called at [/app/public/core-dev/src/wp-includes/plugin.php:189]
#16 apply_filters() called at [/app/public/core-dev/src/wp-includes/option.php:81]
#17 get_option() called at [/app/public/core-dev/src/wp-includes/theme.php:315]
#18 get_template() called at [/app/public/core-dev/src/wp-includes/theme.php:326]
#19 get_template_directory() called at [/app/public/core-dev/src/wp-includes/default-constants.php:393]
#20 wp_templating_constants() called at [/app/public/core-dev/src/wp-settings.php:481]
#21 require_once(/app/public/core-dev/src/wp-settings.php) called at [/app/public/core-dev/wp-config.php:3]
#22 require_once(/app/public/core-dev/wp-config.php) called at [/app/public/core-dev/src/wp-load.php:42]
#23 require_once(/app/public/core-dev/src/wp-load.php) called at [/app/public/core-dev/src/wp-blog-header.php:13]
#24 require(/app/public/core-dev/src/wp-blog-header.php) called at [/app/public/index.php:24]
```

This being the case, an AMP plugin file source array rather than empty? For example:

```php
		if ( $recursion_protection ) {
			return $this->get_file_source_array(
				self::TYPE_PLUGIN,
				'amp',
				substr( $file, strlen( AMP__DIR__ ) + 1 )
			);
		}
```

Granted, this is probably overkill.